File: 815-bus-routes/815-bus-routes.go
// stop to route mapping

func numBusesToDestination(routes [][]int, source int, target int) int {
    graph := buildGraph(routes)
    queue := []int{source}
    visitedStops := make(map[int]struct{})
    visitedRoutes := make(map[int]struct{})
    var numBuses int
    var busStop int
    for len(queue) != 0 {
        for currLen := len(queue); currLen != 0; currLen-- {
            busStop, queue = queue[0], queue[1:]
            if busStop == target {
                return numBuses
            }
            for routeId := range graph[busStop] {
                if _, visitedRoute := visitedRoutes[routeId]; visitedRoute {
                    continue
                }
                for _, nextBusStop := range routes[routeId] {
                    if _, visitedBusStop := visitedStops[nextBusStop]; visitedBusStop {
                        continue
                    }
                    queue = append(queue, nextBusStop)
                    visitedStops[nextBusStop] = struct{}{}
                }
                visitedRoutes[routeId] = struct{}{}
            } 
        }
        numBuses++
    }
    return -1
}

func buildGraph(routes [][]int) map[int]map[int]struct{} {
    graph := make(map[int]map[int]struct{})
    for routeId, busStops := range routes {
        for _, stop := range busStops {
            if _, exists := graph[stop]; !exists {
                graph[stop] = make(map[int]struct{})
            }
            graph[stop][routeId] = struct{}{} 
        }
    }
    return graph
}

File: 815-bus-routes/815-bus-routes.py
"""
time: O(m*n)
space: O(m*n)
"""


class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        graph = defaultdict(set)
        for i, route in enumerate(routes):
            for bus_stop in route:
                graph[bus_stop].add(i)
        num_routes = 0
        queue = collections.deque([source])
        visited_stops = set()
        visited_routes = set()
        visited_stops.add(source)
        while len(queue) != 0:
            for _ in range(len(queue)):
                bus_stop = queue.popleft()
                if bus_stop == target:
                    return num_routes
                for route_id in graph[bus_stop]:
                    if route_id not in visited_routes:
                        for next_stop in routes[route_id]:
                            if next_stop not in visited_stops:
                                queue.append(next_stop)
                                visited_stops.add(next_stop)
                        visited_routes.add(route_id)
            num_routes+=1
                
        return -1